### PR TITLE
Forbid manual updates to email, handle, display for SCIM-managed users

### PR DESCRIPTION
**Pull request by smatting**
Implements https://wearezeta.atlassian.net/browse/SQSERVICES-171

### CODE REVIEW

File: services/brig/src/Brig/API/User.hs [hunk 1]
```diff
@@ -394,13 +400,20 @@ checkRestrictedUserCreation new = do
...
+updateUser :: UserId -> Maybe ConnId -> UserUpdate -> AllowSCIMUpdates -> ExceptT UpdateProfileError AppIO ()
+updateUser uid mconn uu allowScim = do
+  for_ (uupName uu) $ \newName -> do
+    mbUser <- lift $ Data.lookupUser WithPendingInvitations uid
+    user <- maybe (throwE (ProfileNotFound uid)) pure mbUser
+    when
+      ( userManagedBy user == ManagedByScim
+          && userDisplayName user /= newName
+          && allowScim == ForbidSCIMUpdates
+      )
```
Comment (fisx) [inline services/brig/src/Brig/API/User.hs:412]:
```suggestion
    unless
      ( userManagedBy user /= ManagedByScim
          || userDisplayName user == newName
          || allowScim == AllowSCIMUpdates
      )
```

(it's a matter of taste, but I find this easier to read.)

File: services/brig/src/Brig/API/User.hs [hunk 2]
```diff
@@ -421,14 +434,20 @@ changeManagedBy uid conn (ManagedByUpdate mb) = do
...
     throwE ChangeHandleInvalid
   usr <- lift $ Data.lookupUser WithPendingInvitations uid
   case usr of
     Nothing -> throwE ChangeHandleNoIdentity
-    Just u -> claim u
+    Just u -> do
+      when
+        ( userManagedBy u == ManagedByScim
+            && allowScim == ForbidSCIMUpdates
+        )
```
Comment (fisx) [inline services/brig/src/Brig/API/User.hs:448]:
```suggestion
      unless
        ( userManagedBy u /= ManagedByScim
            || old handle == new handle  -- pseudo-code
            || allowScim /= ForbidSCIMUpdates
        )
```

File: services/brig/src/Brig/API/User.hs [hunk 3]
```diff
@@ -525,6 +544,8 @@ changeEmail u email = do
     -- The user already has an email address and the new one is exactly the same
     Just current | current == em -> return ChangeEmailIdempotent
     _ -> do
+      when (userManagedBy usr == ManagedByScim && allowScim == ForbidSCIMUpdates) $
```
Comment (fisx) [inline services/brig/src/Brig/API/User.hs:547]: like above, you *could* change this to `unless ...`.

**Review by fisx — CHANGES_REQUESTED** [review]

**Comment by fisx** [comment]
```
Failures:

  test-integration/Test/Spar/Scim/UserSpec.hs:1594:23: 
  1) Spar.Scim.User, email validation, enabled in team, gives user email
       uncaught exception: ErrorCall
       Assertions failed:
        1: 201 =/= 500
       
       Response was:
       
       Response {responseStatus = Status {statusCode = 500, statusMessage = "Internal Server Error"}, responseVersion = HTTP/1.1, responseHeaders = [("Transfer-Encoding","chunked"),("Date","Fri, 08 Jan 2021 11:26:14 GMT"),("Server","Warp/3.3.13"),("Content-Type","application/json")], responseBody = Just "{\"code\":500,\"message\":\"{\\\"status\\\":\\\"500\\\",\\\"schemas\\\":[\\\"urn:ietf:params:scim:api:messages:2.0:Error\\\"],\\\"detail\\\":\\\"{\\\\\\\"code\\\\\\\":403,\\\\\\\"message\\\\\\\":\\\\\\\"Updating \\\\\\\\\\\\\\\"email\\\\\\\\\\\\\\\" is not allowed, because it is managed by SCIM\\\\\\\",\\\\\\\"label\\\\\\\":\\\\\\\"property-managed-by-scim\\\\\\\"}\\\"}\",\"label\":\"server-error\"}", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}
       CallStack (from HasCallStack):
         error, called at src/Bilge/Assert.hs:88:5 in bilge-0.22.0-LWSwcj5o3J5utdi7oQriS:Bilge.Assert
         <!!, called at test-integration/Util/Scim.hs:191:5 in main:Util.Scim
         createUser, called at test-integration/Test/Spar/Scim/UserSpec.hs:1579:29 in main:Test.Spar.Scim.UserSpec
         setup, called at test-integration/Test/Spar/Scim/UserSpec.hs:1594:23 in main:Test.Spar.Scim.UserSpec

  To rerun use: --match "/Spar.Scim.User/email validation/enabled in team/gives user email/"

Randomized with seed 1959313722
```

I think when spar is calling an internal brig end-point here to trigger email validation, and that end-point needs to call `changeEmail` of `changeSelfEmail` with `AllowSCIMUpdates`, not `ForbidSCIMUpdates`.

You should check if there are any other calls to this end-point and make sure that it's always `AllowSCIMUpdates`, or else make it a query parameter (mandatory if you want to be more certain the integration tests capture anything you've missed).

File: services/brig/brig.cabal [hunk 4]
```diff
@@ -1,6 +1,6 @@
 cabal-version: 1.12
 
--- This file has been generated from package.yaml by hpack version 0.33.0.
+-- This file has been generated from package.yaml by hpack version 0.34.3.
```
[the remark this answers is not in the document]
Comment (smatting) [reply services/brig/brig.cabal:3]: Thanks! This file was accedentialy overwritten bei my HLS. I'm using separate stack work dirs from preventing this now

File: services/spar/test-integration/Test/Spar/Scim/UserSpec.hs [hunk 5]
```diff
@@ -1594,3 +1597,42 @@ specEmailValidation = do
...
+      (tok, _) <- registerIdPAndScimToken
+      user <- randomScimUser
+      storedUser <- createUser tok user
+      let uid = Scim.id . Scim.thing $ storedUser
+
+      do
+        email <- randomEmail
+        call $
+          changeEmailBrig brig uid email !!! do
+            (fmap Wai.label . responseJsonEither @Wai.Error) === const (Right "property-managed-by-scim")
```
[the remark this answers is not in the document]
Comment (smatting) [reply services/spar/test-integration/Test/Spar/Scim/UserSpec.hs:1617]: Renamed to `managed-by-scim`

File: services/brig/src/Brig/API/User.hs (hunk 2 above)
Comment (smatting) [reply services/brig/src/Brig/API/User.hs:448]: done

File: services/brig/src/Brig/API/User.hs (hunk 3 above)
Comment (smatting) [reply services/brig/src/Brig/API/User.hs:547]: done

**Comment by smatting** [comment]
Fixed the failing test by allow scim updates on the internal endpoint `/i/self/email` (I failed to see this before).
I think this is a safe choice, because only public endpoints are have a change in behaviour now.

**Pull request merged**